packages/editor/src/client/atoms/poll/pollExternalVotes.tsx

```diff
@@ -0,0 +1,265 @@
+  const [sourceToDelete, setSourceToDelete] = useState<PollExternalVoteSource | undefined>(
+    undefined
+  )
+  const [openModal, setOpenModal] = useState<boolean>(false)
+  const [createSourceMutation, {loading}] = useCreatePollExternalVoteSourceMutation({
+    fetchPolicy: 'no-cache'
+  })
+  const [deletePollMutation] = useDeletePollExternalVoteSourceMutation({
+    fetchPolicy: 'no-cache'
+  })
+
+  const onErrorToast = (error: ApolloError) => {
+    toaster.push(
+      <Message type="error" showIcon closable duration={3000}>
+        {error.message}
+      </Message>
+    )
+  }
+
+  async function createPoll() {
+    if (!newSource) {
+      toaster.push(
+        <Message showIcon type="error" closable duration={3000}>
+          {t('pollExternalVotes.emptySource')}
+        </Message>
+      )
+      return
+    }
+    if (!poll) {
+      toaster.push(
+        <Message showIcon type="error" closable duration={3000}>
+          {t('pollExternalVotes.noPollAvailable')}
+        </Message>
+      )
+      return
+    }
+    // first save current poll state
+    await savePoll()
+
+    await createSourceMutation({
+      variables: {
+        pollId: poll.id,
+        source: newSource
+      },
+      onError: onErrorToast
+    })
+    setNewSource(undefined)
+    await onExternalSourceCreated()
+  }
+
+  async function deletePoll() {
+    const id = sourceToDelete?.id
+    if (!id) {
+      return
+    }
+    // first save current poll state
+    await savePoll()
+    await deletePollMutation({
+      variables: {
+        deletePollExternalVoteSourceId: id
+      }
+    })
+    setSourceToDelete(undefined)
+    setOpenModal(false)
+    await onExternalSourceDeleted()
+  }
+
+  /**
+   * set a new amount of an external vote source. used to update the parents vote source prop.
+   * @param answer
+   * @param externalVoteSource
+   * @param newAmount
+   */
+  function changeSource(
+    answer: PollAnswerWithVoteCount,
+    externalVoteSource: PollExternalVoteSource,
+    newAmount: string | number
+  ) {
+    const voteSources = poll?.externalVoteSources
+    if (!voteSources) {
+      return
+    }
+    // deep clone
+    const newVoteSources: PollExternalVoteSource[] = [...voteSources]
+
+    const sourceIndex = newVoteSources.findIndex(
+      (tmpVoteSource: PollExternalVoteSource) => tmpVoteSource.source === externalVoteSource.source
+    )
+    if (sourceIndex < 0) {
+      return
+    }
+
+    const source = newVoteSources[sourceIndex]
+    if (!source?.voteAmounts) {
+      return
+    }
+
+    const voteIndex = source.voteAmounts?.findIndex(
+      (tmpVote: PollExternalVote) => tmpVote.answerId === answer.id
+    )
+    if (voteIndex < 0) {
+      return
+    }
+    source.voteAmounts[voteIndex].amount =
+      typeof newAmount === 'string' ? parseInt(newAmount) : newAmount
+
+    onExternalSourceChange(newVoteSources)
+  }
+
+  /**
+   * UI HELPER FUNCTIONS
+   */
+  function tableView() {
```

Comment:
Could we make this pure (or even better a component) and move it out of the component?

Reply:
These functions are closely connected to it's component. It would not make any sense to export them as a new component. Rather it would generate boilerplate and complicate things up.

I just split this code into three parts, so that the code structure is more clear and clean and one line of code within a function has the same level of complexity.
![image](https://user-images.githubusercontent.com/70939054/188429681-ca89fefb-b09a-44c6-a726-8c57329ba148.png)